velox/functions/prestosql/MapSubset.h

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/expression/ComplexViewTypes.h"
+#include "velox/functions/Udf.h"
+
+namespace facebook::velox::functions {
+
+// Fast path for constant primitive type keys: map_subset(m, array[1, 2, 3]).
+#define VELOX_MAP_SUBSET_PRIMITIVE_FUNCTION(TName, TType)           \
+  template <typename TExec>                                         \
+  struct MapSubset##TName##Function {                               \
+    VELOX_DEFINE_FUNCTION_TYPES(TExec);                             \
+                                                                    \
+    void initialize(                                                \
+        const core::QueryConfig& /*config*/,                        \
+        const arg_type<Map<TType, Generic<T1>>>* /*inputMap*/,      \
+        const arg_type<Array<TType>>* keys) {                       \
+      if (keys != nullptr) {                                        \
+        constantSearchKeys_ = true;                                 \
+        initializeSearchKeys(*keys);                                \
+      }                                                             \
+    }                                                               \
+                                                                    \
+    void call(                                                      \
+        out_type<Map<TType, Generic<T1>>>& out,                     \
+        const arg_type<Map<TType, Generic<T1>>>& inputMap,          \
+        const arg_type<Array<TType>>& keys) {                       \
+      if (keys.empty()) {                                           \
+        return;                                                     \
+      }                                                             \
+                                                                    \
+      if (!constantSearchKeys_) {                                   \
```

Comment:
What's the difference of constantSearchKeys_ and searchKeys_? What are the keys passed in initialize() and call()? Thanks!

Reply:
'initialize' is called once per thread of execution. It receives constant arguments. 'call' is called once per row. It receives per-row arguments.

For example, 

```
map_subset(m, array[1,2,3])
```

In this case, 'initialize' will receive array[1, 2, 3], then all 'call' invocations will receive array[1,2,3] again.

```
initialize(...array[1,2,3])
call(m1, array[1,2,3])
call(m2, array[1,2,3])
call(m3, array[1,2,3])
...
```

Now,

```
map_subset(m, a)
```

In this case, assuming 'a' is a column or an expression of type array, 'initialize' will receive a nullptr (argument is not constant), but 'call' will receive the actual values of the array.

```
initialize(...nullptr)
call(m1, a1);
call(m2, a2);
call(m3, a3);
...
```

All uses of map_subset I saw specify constant array of keys. Hence, we can optimize by building a set of keys once (in 'initialize') and avoid creating a set for every row.

Hope this helps.

Reply:
I see. Thanks for the explanation. It might be good to comment in the code say constantSearchKeys_ is set if the search keys is passed in initialize() call to save passing search keys in each call invocation for optimization. Thanks!